src/dparse/lexer.d

```diff
@@ -2001,9 +1968,297 @@ unittest
+            if (lines[i][0] != commentChar)
+                break;
+            else if (i == hi-1)
+                allDecorated = true;
+        }
+        if (!allDecorated)
+            return;
+
+        const size_t indexToChange = (lines[lo][0] == commentChar) ? 0 : 1;
+        foreach (const i; lo .. hi)
+            lines[i][indexToChange] = ' ';
+    }
+
+    void stripLeft() @safe @nogc pure nothrow
+    {
+        foreach (const i; 0 .. lines[0].length)
+            if (!(lines[0][i]).isWhite)
+        {
+            lines[0] = lines[0][i..$];
+            break;
+        }
+        if (numLines == 1)
+            return;
+        while (true)
+        {
+            bool processColumn;
+            foreach (const i; 1 .. lastLineInBlockPlusOne)
+            {
+                if (lines[i].length == 0)
+                    continue;
+                if (!(lines[i][0]).isWhite)
+                    return;
+                processColumn = true;
+            }
+            if (!processColumn)
+                return;
+            foreach (const i; 1 .. lastLineInBlockPlusOne)
+            {
+                if (lines[i].length == 0)
+                    continue;
+                lines[i] = lines[i][1..$];
+            }
+        }
+    }
+
+    void stripRight() @safe @nogc pure nothrow
+    {
+        foreach (const i; 0 .. numLines)
+        {
+            if (lines[i].length == 0)
+                continue;
+            if ((lines[i][$-1]).isWhite)
+            {
+                ptrdiff_t firstWhite = lines[i].length;
+                while (firstWhite > 0 && (lines[i][firstWhite-1]).isWhite)
+                    firstWhite--;
+                lines[i] = lines[i][0..firstWhite];
+            }
+        }
+    }
+
+    void run() @safe @nogc pure nothrow
+    {
+        stripIndent();
+        processFirstLine();
+        processLastLine();
+        unDecorate();
+        stripLeft();
+        stripRight();
+    }
+
+public:
+
+    this(CharType[] text) @safe pure nothrow
+    {
+        commentChar = text[1];
+        size_t startIndex, i;
+
+        void storeLine(size_t endIndexPlusOne)
+        {
+            static if (isMutable!CharType)
+                lines[numLines++] = text[startIndex..endIndexPlusOne];
+            else
+                lines[numLines++] = text[startIndex..endIndexPlusOne].dup;
+        }
+
+        while (true)
+        {
+            if (i == text.length - 1)
+            {
+                storeLine(text.length);
+                break;
+            }
+            if (text[i] == '\n')
+            {
+                storeLine(i);
+                startIndex = i + 1;
+            }
+            else if (text[i .. i+2] == "\r\n")
+            {
+                storeLine(i);
+                i++;
+                startIndex = i + 1;
+            }
+            i++;
+            if (numLines == MaxLines)
+                break;
+        }
+    }
+
+    void process(T)(ref T outbuffer)
```

Comment:
I can add a constraint but since the whole struct is private i didin't think this was necessary.
D template constraints are over used IMO. And here there's no overload to disambiguate.